internal/expr/bindtypes.go

```diff
@@ -49,64 +49,93 @@ func (pe *ParsedExpr) BindTypes(args ...any) (tbe *TypeBoundExpr, err error) {
 	// Bind types to each expression.
 	var typedExprs TypeBoundExpr
 	outputUsed := map[typeinfo.Output]bool{}
-	var te any
 	for _, expr := range pe.exprs {
-		switch e := expr.(type) {
-		case *inputExpr:
-			te, err = bindInputTypes(e, argInfo)
-			if err != nil {
-				return nil, err
-			}
-		case *outputExpr:
-			toe, err := bindOutputTypes(e, argInfo)
-			if err != nil {
-				return nil, err
-			}
+		te, err := expr.bindTypes(argInfo)
+		if err != nil {
+			return nil, err
+		}
 
+		if toe, ok := te.(*typedOutputExpr); ok {
 			for _, oc := range toe.outputColumns {
 				if ok := outputUsed[oc.output]; ok {
 					return nil, fmt.Errorf("%s appears more than once in output expressions", oc.output.String())
 				}
 				outputUsed[oc.output] = true
 			}
-			te = toe
-		case *bypass:
-			te = e
-		default:
-			return nil, fmt.Errorf("internal error: unknown query expr type %T", expr)
 		}
 		typedExprs = append(typedExprs, te)
 	}
 
 	return &typedExprs, nil
 }
 
-// bindInputTypes binds the input expression to a query type and returns a typed
-// input expression.
-func bindInputTypes(e *inputExpr, argInfo typeinfo.ArgInfo) (te *typedInputExpr, err error) {
-	defer func() {
-		if err != nil {
-			err = fmt.Errorf("input expression: %s: %s", err, e.raw)
-		}
-	}()
+// expression represents a parsed node of the SQLair query's AST.
+type expression interface {
+	// String returns a text representation for debugging and testing purposes.
+	String() string
 
-	var input typeinfo.Input
-	switch a := e.sourceType.(type) {
-	case memberAccessor:
-		input, err = argInfo.InputMember(a.typeName, a.memberName)
-		if err != nil {
-			return nil, err
-		}
-	case sliceAccessor:
-		return nil, fmt.Errorf("slice support not implemented")
+	// bindTypes binds the types to the expression to generate either a
+	// *typedInputExpr or *typedOutputExpr.
+	bindTypes(typeinfo.ArgInfo) (any, error)
+}
+
+// memberInputExpr is an input expression of the form "$Type.member" which
+// represents a query parameter contained in a member of a type.
+type memberInputExpr struct {
+	raw string
+	ma  memberAccessor
```

Comment:
Maybe we can go one step further because it would be nice to flatten this and remove the `memberInputExpr` that is only a wrapper. Could we implement  `bindTypes` on `memberAccessor` directly?

The challenging bit would be the separation between input and output expressions and making that clear.

Reply:
We could, but I think this would get a bit confusing when the INSERT expressions are added. In their current implementation the insert input expression type has a slice of `memberAccessor`. We could add `memberName` and `typeName` directly to `memberInputExpr` rather than having the `memberAccessor` as a field but I think when the INSERT PR is added this will make more sense.

Of course if we change how it is implemented in the INSERT PR then it makes sense to change it here.

Reply:
I see, then we can go ahead with this PR and then discuss it further in the INSERT one. But taking into account what you just mentioned I think the wrapper probably unavoidable.